# BDDéšå±¤æ§‹é€ çµ±ä¸€ãƒ«ãƒ¼ãƒ«

ã“ã®ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã¯ã€ag-loggerãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã«ãŠã‘ã‚‹BDDï¼ˆBehavior-Driven Developmentï¼‰ãƒ†ã‚¹ãƒˆéšå±¤æ§‹é€ ã®çµ±ä¸€ãƒ«ãƒ¼ãƒ«ã‚’å®šç¾©ã—ã¾ã™ã€‚

## ğŸ¯ åŸºæœ¬åŸå‰‡

### 1. 3éšå±¤BDDæ§‹é€ ã®å³æ ¼éµå®ˆ

**ã™ã¹ã¦ã®ãƒ†ã‚¹ãƒˆãƒ•ã‚¡ã‚¤ãƒ«ã¯ä»¥ä¸‹ã®3éšå±¤æ§‹é€ ã«çµ±ä¸€ã™ã‚‹:**

```typescript
describe('[Level 1: Suite]', () => {
  describe('[Level 2: Context]', () => {
    it('[Level 3: Specification]', () => {
      // ãƒ†ã‚¹ãƒˆå®Ÿè£…
    });
  });
});
```

### 2. 4éšå±¤ä»¥ä¸Šã®ãƒã‚¹ãƒˆå®Œå…¨ç¦æ­¢

- **æœ€é«˜å„ªå…ˆäº‹é …**: 4éšå±¤ä»¥ä¸Šã®ãƒã‚¹ãƒˆã¯çµ¶å¯¾ã«ç¦æ­¢
- **æ¤œå‡ºæ–¹æ³•**: MCPãƒ„ãƒ¼ãƒ«ã§ãƒ‘ã‚¿ãƒ¼ãƒ³æ¤œç´¢å®Ÿè¡Œ
- **ä¿®æ­£æ–¹æ³•**: 3éšå±¤ä»¥ä¸‹ã«å†æ§‹æˆ

```bash
# 4éšå±¤ä»¥ä¸Šæ¤œå‡ºã‚³ãƒãƒ³ãƒ‰
mcp__serena-mcp__search_for_pattern --substring_pattern "describe.*describe.*describe.*describe" --paths_include_glob "*.spec.ts"
```

## ğŸ“ éšå±¤åˆ¥æ§‹é€ ãƒ«ãƒ¼ãƒ«

### Level 1: Suiteï¼ˆæœ€ä¸Šä½ï¼‰

**3ã¤ã®ãƒ‘ã‚¿ãƒ¼ãƒ³ã‹ã‚‰é¸æŠ:**

#### ãƒ‘ã‚¿ãƒ¼ãƒ³A: Givené–‹å§‹ï¼ˆå‰ææ¡ä»¶ãƒ™ãƒ¼ã‚¹ï¼‰

```typescript
describe('Given: [å‰ææ¡ä»¶ã®è©³ç´°è¨˜è¿°]', () => {
```

#### ãƒ‘ã‚¿ãƒ¼ãƒ³B: Featureé–‹å§‹ï¼ˆæ©Ÿèƒ½å˜ä½ãƒ†ã‚¹ãƒˆï¼‰

```typescript
describe('Feature: [æ©Ÿèƒ½åãƒ»ç‰¹å¾´ã®è¨˜è¿°]', () => {
```

#### ãƒ‘ã‚¿ãƒ¼ãƒ³C: Thené–‹å§‹ï¼ˆæœŸå¾…çµæœè©³ç´°ï¼‰

```typescript
describe('Then: [æ­£å¸¸]/[ç•°å¸¸]/[ã‚¨ãƒƒã‚¸ã‚±ãƒ¼ã‚¹] - [è©³ç´°ãªæœŸå¾…çµæœ]', () => {
```

### Level 2: Contextï¼ˆå®Ÿè¡Œæ¡ä»¶ï¼‰

**Whené–‹å§‹ã§çµ±ä¸€:**

```typescript
describe('When: [å…·ä½“çš„ãªæ“ä½œãƒ»å®Ÿè¡Œå†…å®¹]', () => {
```

### Level 3: Specificationï¼ˆæ¤œè¨¼å†…å®¹ï¼‰

**ã‚±ãƒ¼ã‚¹ç¨®åˆ¥ã‚¿ã‚°å¿…é ˆ + Thenå½¢å¼:**

```typescript
it('Then: [æ­£å¸¸]/[ç•°å¸¸]/[ã‚¨ãƒƒã‚¸ã‚±ãƒ¼ã‚¹] - should [æœŸå¾…ã•ã‚Œã‚‹å‹•ä½œ]', () => {
```

## ğŸ·ï¸ ã‚±ãƒ¼ã‚¹ç¨®åˆ¥ã‚¿ã‚°ï¼ˆå¿…é ˆï¼‰

### åˆ†é¡ã‚¿ã‚°ã®å®šç¾©

| ã‚¿ã‚°               | ç”¨é€”        | èª¬æ˜                         |
| ------------------ | ----------- | ---------------------------- |
| **[æ­£å¸¸]**         | Happy Path  | æœŸå¾…å…¥åŠ›ã§ã®æ­£å¸¸å‹•ä½œç¢ºèª     |
| **[ç•°å¸¸]**         | Error Cases | ã‚¨ãƒ©ãƒ¼ãƒ»ä¾‹å¤–å‡¦ç†ã®ç¢ºèª       |
| **[ã‚¨ãƒƒã‚¸ã‚±ãƒ¼ã‚¹]** | Edge Cases  | å¢ƒç•Œå€¤ãƒ»ç‰¹æ®Šæ¡ä»¶ã§ã®å‹•ä½œç¢ºèª |

### é©ç”¨ä¾‹

```typescript
describe('Feature: User authentication', () => {
  describe('When: logging in with credentials', () => {
    it('Then: [æ­£å¸¸] - should return valid token for correct credentials', () => {
      // æ­£å¸¸ç³»ãƒ†ã‚¹ãƒˆ
    });

    it('Then: [ç•°å¸¸] - should throw error for invalid password', () => {
      // ç•°å¸¸ç³»ãƒ†ã‚¹ãƒˆ
    });

    it('Then: [ã‚¨ãƒƒã‚¸ã‚±ãƒ¼ã‚¹] - should handle empty username gracefully', () => {
      // ã‚¨ãƒƒã‚¸ã‚±ãƒ¼ã‚¹ãƒ†ã‚¹ãƒˆ
    });
  });
});
```

## âš ï¸ ç¦æ­¢äº‹é …

### 1. TOPãƒ¬ãƒ™ãƒ«ã‚¯ãƒ©ã‚¹ådescribeå‰Šé™¤

```typescript
// âŒ ç¦æ­¢: ã‚¯ãƒ©ã‚¹åãŒãã®ã¾ã¾describe
describe('AgLogger', () => {

// âœ… æ¨å¥¨: æ©Ÿèƒ½ãƒ»å‰ææ¡ä»¶ãƒ™ãƒ¼ã‚¹
describe('Feature: AgLogger instance management', () => {
```

### 2. 4éšå±¤ä»¥ä¸Šã®ãƒã‚¹ãƒˆ

```typescript
// âŒ ç¦æ­¢: 4éšå±¤
describe('Suite', () => {
  describe('Context1', () => {
    describe('Context2', () => { // 3éšå±¤
      describe('Context3', () => { // 4éšå±¤ â† ç¦æ­¢
        it('test', () => {});
      });
    });
  });
});
```

### 3. ã‚±ãƒ¼ã‚¹ç¨®åˆ¥ã‚¿ã‚°ã®çœç•¥

```typescript
// âŒ ç¦æ­¢: ã‚¿ã‚°ãªã—
it('should return user data', () => {});

// âœ… æ¨å¥¨: ã‚¿ã‚°ä»˜ã
it('Then: [æ­£å¸¸] - should return user data', () => {});
```

## ğŸ”§ ä¿®æ­£å„ªå…ˆé †ä½

### æœ€é«˜å„ªå…ˆï¼ˆCriticalï¼‰

1. **4éšå±¤ä»¥ä¸Šã®ãƒã‚¹ãƒˆå•é¡Œä¿®æ­£**
   - å³åº§ã«3éšå±¤ä»¥ä¸‹ã«å†æ§‹æˆ
   - ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆå…¨ä½“ã®æ§‹é€ æ•´åˆæ€§ã«å½±éŸ¿

### é«˜å„ªå…ˆï¼ˆHighï¼‰

2. **TOPãƒ¬ãƒ™ãƒ«ã‚¯ãƒ©ã‚¹ådescribeã®å‰Šé™¤**
   - æ©Ÿèƒ½ãƒ™ãƒ¼ã‚¹ã®å‘½åã«å¤‰æ›´
   - ãƒ†ã‚¹ãƒˆã®æ„å›³ã‚’æ˜ç¢ºåŒ–

### ä¸­å„ªå…ˆï¼ˆMediumï¼‰

3. **å…¨itæ–‡ã§Then:ã‚¿ã‚°å½¢å¼ã®é©ç”¨**
   - ã‚±ãƒ¼ã‚¹ç¨®åˆ¥ã‚¿ã‚°ã®è¿½åŠ 
   - ä¸€è²«æ€§å‘ä¸Š

### ä½å„ªå…ˆï¼ˆLowï¼‰

4. **Given/Featureé–‹å§‹ãƒ‘ã‚¿ãƒ¼ãƒ³ã®é©åˆ‡é¸æŠ**
   - æœ€é©ãªãƒ‘ã‚¿ãƒ¼ãƒ³é¸æŠ
   - å¯èª­æ€§å‘ä¸Š

## ğŸ› ï¸ MCPãƒ„ãƒ¼ãƒ«æ´»ç”¨

### BDDå½¢å¼èª¿æŸ»

```bash
# Given/When/Thenå½¢å¼ã®èª¿æŸ»
mcp__serena-mcp__search_for_pattern --substring_pattern "Given:|When:|Then:" --paths_include_glob "*.spec.ts"

# ãƒ†ã‚¹ãƒˆãƒ•ã‚¡ã‚¤ãƒ«æ§‹é€ åˆ†æ
mcp__serena-mcp__get_symbols_overview --relative_path "src/__tests__/[å¯¾è±¡ãƒ•ã‚¡ã‚¤ãƒ«]"
```

### å•é¡Œãƒ‘ã‚¿ãƒ¼ãƒ³æ¤œå‡º

```bash
# 4éšå±¤ä»¥ä¸Šæ¤œå‡º
mcp__serena-mcp__search_for_pattern --substring_pattern "describe.*describe.*describe.*describe" --paths_include_glob "*.spec.ts"

# ã‚¯ãƒ©ã‚¹ådescribeæ¤œå‡º
mcp__serena-mcp__search_for_pattern --substring_pattern "describe\\('\\w+'" --paths_include_glob "*.spec.ts"
```

## ğŸ“Š é©ç”¨ç¯„å›²

### å¯¾è±¡ãƒ•ã‚¡ã‚¤ãƒ«

- **Unit Tests**: `src/__tests__/**/*.spec.ts`
- **Functional Tests**: `src/__tests__/functional/**/*.spec.ts`
- **Integration Tests**: `tests/integration/**/*.spec.ts`
- **E2E Tests**: `tests/e2e/**/*.spec.ts`

### é©ç”¨çŠ¶æ³

- **Phase 1**: E2Eãƒ†ã‚¹ãƒˆï¼ˆ8ãƒ•ã‚¡ã‚¤ãƒ«ï¼‰- âœ… å®Œäº†
- **Phase 2-4**: Integration/Functional/Unitï¼ˆ36ãƒ•ã‚¡ã‚¤ãƒ«ï¼‰- ğŸ”„ é€²è¡Œä¸­

---

**é‡è¦**: ã“ã®ãƒ«ãƒ¼ãƒ«ã¯ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆå…¨ä½“ã®ãƒ†ã‚¹ãƒˆå“è³ªãƒ»ä¿å®ˆæ€§å‘ä¸Šã®ãŸã‚ã€**ä¾‹å¤–ãªãéµå®ˆ**ã™ã‚‹ã“ã¨ã€‚

**å‚ç…§**: [todo.mdä½œæ¥­é€²æ—](../../packages/@aglabo/agla-logger/todo.md)
